File: agi_scripts/call_routing.js
#!/usr/bin/env node

/**
 * AGI Script สำหรับจัดการ call routing ผ่าน Google Speech API
 * ใช้ Node.js และ AEP (Asterisk External Protocol)
 */

const fs = require('fs');
const path = require('path');
const temp = require('temp');
const Database = require('../config/database');
const GoogleSpeechService = require('../config/google_speech');

class CallRoutingAGI {
    constructor() {
        this.speechService = new GoogleSpeechService();
        this.database = new Database();
        this.tempDir = temp.track();
    }

    log(message) {
        /**เขียน log ไปยัง Asterisk*/
        console.log(`VERBOSE "${message}" 1`);
        process.stdout.write(`VERBOSE "${message}" 1\n`);
    }

    async agiCommand(command) {
        /**ส่งคำสั่งไปยัง Asterisk*/
        console.log(command);
        process.stdout.write(`${command}\n`);
        
        return new Promise((resolve) => {
            process.stdin.once('data', (data) => {
                resolve(data.toString().trim());
            });
        });
    }

    async playAudioFile(filename) {
        /**เล่นไฟล์เสียง*/
        return await this.agiCommand(`STREAM FILE ${filename} ""`);
    }

    async recordAudio(filename, duration = 5, silence = 2) {
        /**บันทึกเสียงจากผู้โทร*/
        return await this.agiCommand(`RECORD FILE ${filename} wav ${duration} ${silence}`);
    }

    async sayText(text) {
        /**พูดข้อความผ่าน TTS*/
        try {
            // แปลงข้อความเป็นเสียง
            const audioData = await this.speechService.textToSpeech(text);
            if (!audioData) {
                return false;
            }

            // สร้างไฟล์เสียงชั่วคราว
            const tempFile = temp.path({ suffix: '.wav' });
            await this.speechService.createAudioFile(audioData, tempFile);

            // เล่นไฟล์เสียง
            const result = await this.playAudioFile(tempFile);

            // ลบไฟล์ชั่วคราว
            try {
                fs.unlinkSync(tempFile);
            } catch (err) {
                this.log(`Warning: Could not delete temp file ${tempFile}: ${err.message}`);
            }

            return result;
        } catch (error) {
            this.log(`Error in sayText: ${error.message}`);
            return false;
        }
    }

    async getSpeechInput(promptText, maxAttempts = 3) {
        /**รับข้อมูลเสียงจากผู้โทร*/
        for (let attempt = 0; attempt < maxAttempts; attempt++) {
            // พูดข้อความถาม
            await this.sayText(promptText);

            // บันทึกเสียง
            const tempFile = temp.path({ suffix: '.wav' });
            this.log(`Recording attempt ${attempt + 1}`);

            const recordResult = await this.recordAudio(tempFile, 10, 3);

            if (recordResult && recordResult.includes('200')) {
                // อ่านไฟล์เสียง
                try {
                    const audioData = await this.speechService.readAudioFile(tempFile);

                    // แปลงเสียงเป็นข้อความ
                    const text = await this.speechService.speechToText(audioData);

                    // ลบไฟล์ชั่วคราว
                    try {
                        fs.unlinkSync(tempFile);
                    } catch (err) {
                        this.log(`Warning: Could not delete temp file ${tempFile}: ${err.message}`);
                    }

                    if (text) {
                        this.log(`Recognized text: ${text}`);
                        return text;
                    } else {
                        this.log('No speech recognized');
                        if (attempt < maxAttempts - 1) {
                            await this.sayText('ขออภัย ไม่ได้ยินเสียง กรุณาพูดอีกครั้ง');
                        }
                    }
                } catch (error) {
                    this.log(`Error processing audio: ${error.message}`);
                    try {
                        fs.unlinkSync(tempFile);
                    } catch (err) {
                        // Ignore cleanup errors
                    }
                }
            } else {
                this.log('Recording failed');
            }

            if (attempt < maxAttempts - 1) {
                await new Promise(resolve => setTimeout(resolve, 1000));
            }
        }

        return null;
    }

    async confirmBranch(branchName) {
        /**ยืนยันสาขาที่ลูกค้าต้องการติดต่อ*/
        const confirmationText = `คุณต้องการติดต่อสาขา ${branchName} ใช่หรือไม่`;
        await this.sayText(confirmationText);

        // รับคำตอบ
        const response = await this.getSpeechInput('กรุณาพูด ใช่ หรือ ไม่ใช่');

        if (response) {
            // ตรวจสอบคำตอบ
            const responseLower = response.toLowerCase();
            const positiveWords = ['ใช่', 'yes', 'ถูกต้อง', 'ถูก'];
            const negativeWords = ['ไม่ใช่', 'no', 'ไม่ถูก', 'ผิด'];

            if (positiveWords.some(word => responseLower.includes(word))) {
                return true;
            } else if (negativeWords.some(word => responseLower.includes(word))) {
                return false;
            }
        }

        return false;
    }

    async routeCall(extension) {
        /**โอนสายไปยังหมายเลขภายใน*/
        this.log(`Transferring call to extension ${extension}`);
        return await this.agiCommand(`DIAL SIP/${extension},30`);
    }

    async logCallData(callData) {
        /**บันทึกข้อมูลการโทร*/
        try {
            await this.database.logCall(callData);
        } catch (error) {
            this.log(`Error logging call data: ${error.message}`);
        }
    }

    async main() {
        /**ฟังก์ชันหลักของ AGI script*/
        const callData = {
            callerId: process.env.CALLERIDNUM || 'unknown',
            calledNumber: process.env.DNID || 'unknown',
            requestedBranch: null,
            recognizedText: null,
            routedToExtension: null,
            callDuration: 0,
            callStatus: 'answered'
        };

        try {
            this.log('Call routing AGI started');

            // 1. รับข้อมูลสาขาที่ต้องการติดต่อ
            const branchInput = await this.getSpeechInput('กรุณาพูดสาขาที่ต้องการติดต่อ');

            if (!branchInput) {
                await this.sayText('ขออภัย ไม่สามารถรับข้อมูลได้ กรุณาลองใหม่อีกครั้ง');
                callData.callStatus = 'failed';
                await this.logCallData(callData);
                return;
            }

            callData.requestedBranch = branchInput;
            callData.recognizedText = branchInput;
            this.log(`Customer requested branch: ${branchInput}`);

            // 2. ค้นหาสาขาในฐานข้อมูล
            const branchInfo = await this.database.getBranchExtension(branchInput);

            if (!branchInfo) {
                await this.sayText(`ขออภัย ไม่พบสาขา ${branchInput} กรุณาติดต่อเจ้าหน้าที่`);
                callData.callStatus = 'failed';
                await this.logCallData(callData);
                return;
            }

            // 3. ยืนยันสาขา
            if (!(await this.confirmBranch(branchInfo.branch_name))) {
                await this.sayText('กรุณาพูดสาขาที่ต้องการติดต่ออีกครั้ง');
                callData.callStatus = 'failed';
                await this.logCallData(callData);
                return;
            }

            // 4. โอนสาย
            await this.sayText(`กำลังโอนสายไปยังสาขา ${branchInfo.branch_name}`);
            callData.routedToExtension = branchInfo.extension;
            
            const transferResult = await this.routeCall(branchInfo.extension);

            if (transferResult && transferResult.includes('200')) {
                this.log('Call transferred successfully');
                callData.callStatus = 'answered';
            } else {
                await this.sayText('ขออภัย ไม่สามารถโอนสายได้ในขณะนี้ กรุณาติดต่อเจ้าหน้าที่');
                callData.callStatus = 'failed';
            }

            await this.logCallData(callData);

        } catch (error) {
            this.log(`Error in main: ${error.message}`);
            await this.sayText('ขออภัย เกิดข้อผิดพลาดในระบบ กรุณาติดต่อเจ้าหน้าที่');
            callData.callStatus = 'failed';
            await this.logCallData(callData);
        } finally {
            // ปิดการเชื่อมต่อฐานข้อมูล
            await this.database.close();
        }
    }
}

// รัน AGI script
if (require.main === module) {
    const agi = new CallRoutingAGI();
    agi.main().catch(error => {
        console.error('AGI Script Error:', error);
        process.exit(1);
    });
}

module.exports = CallRoutingAGI;

File: agi_scripts/call_routing.py
#!/usr/bin/env python3
"""
AGI Script สำหรับจัดการ call routing ผ่าน Google Speech API
"""

import sys
import os
import tempfile
import time
from config.google_speech import GoogleSpeechService
from config.database import Database

class CallRoutingAGI:
    def __init__(self):
        self.speech_service = GoogleSpeechService()
        self.database = Database()
        
    def log(self, message):
        """เขียน log ไปยัง Asterisk"""
        print(f"VERBOSE \"{message}\" 1", flush=True)
        sys.stdout.flush()
    
    def agi_command(self, command):
        """ส่งคำสั่งไปยัง Asterisk"""
        print(command, flush=True)
        sys.stdout.flush()
        return sys.stdin.readline().strip()
    
    def play_audio_file(self, filename):
        """เล่นไฟล์เสียง"""
        return self.agi_command(f"STREAM FILE {filename} \"\"")
    
    def record_audio(self, filename, duration=5, silence=2):
        """บันทึกเสียงจากผู้โทร"""
        return self.agi_command(f"RECORD FILE {filename} wav {duration} {silence}")
    
    def say_text(self, text):
        """พูดข้อความผ่าน TTS"""
        try:
            # แปลงข้อความเป็นเสียง
            audio_data = self.speech_service.text_to_speech(text)
            if not audio_data:
                return False
            
            # สร้างไฟล์เสียงชั่วคราว
            with tempfile.NamedTemporaryFile(suffix='.wav', delete=False) as temp_file:
                temp_file.write(audio_data)
                temp_filename = temp_file.name
            
            # เล่นไฟล์เสียง
            result = self.play_audio_file(temp_filename)
            
            # ลบไฟล์ชั่วคราว
            os.unlink(temp_filename)
            
            return result
        except Exception as e:
            self.log(f"Error in say_text: {e}")
            return False
    
    def get_speech_input(self, prompt_text, max_attempts=3):
        """รับข้อมูลเสียงจากผู้โทร"""
        for attempt in range(max_attempts):
            # พูดข้อความถาม
            self.say_text(prompt_text)
            
            # บันทึกเสียง
            with tempfile.NamedTemporaryFile(suffix='.wav', delete=False) as temp_file:
                temp_filename = temp_file.name
            
            self.log(f"Recording attempt {attempt + 1}")
            record_result = self.record_audio(temp_filename, duration=10, silence=3)
            
            if record_result and "200" in record_result:
                # อ่านไฟล์เสียง
                try:
                    with open(temp_filename, 'rb') as f:
                        audio_data = f.read()
                    
                    # แปลงเสียงเป็นข้อความ
                    text = self.speech_service.speech_to_text(audio_data)
                    
                    # ลบไฟล์ชั่วคราว
                    os.unlink(temp_filename)
                    
                    if text:
                        self.log(f"Recognized text: {text}")
                        return text
                    else:
                        self.log("No speech recognized")
                        if attempt < max_attempts - 1:
                            self.say_text("ขออภัย ไม่ได้ยินเสียง กรุณาพูดอีกครั้ง")
                except Exception as e:
                    self.log(f"Error processing audio: {e}")
                    if os.path.exists(temp_filename):
                        os.unlink(temp_filename)
            else:
                self.log("Recording failed")
            
            if attempt < max_attempts - 1:
                time.sleep(1)
        
        return None
    
    def confirm_branch(self, branch_name):
        """ยืนยันสาขาที่ลูกค้าต้องการติดต่อ"""
        confirmation_text = f"คุณต้องการติดต่อสาขา {branch_name} ใช่หรือไม่"
        self.say_text(confirmation_text)
        
        # รับคำตอบ
        response = self.get_speech_input("กรุณาพูด ใช่ หรือ ไม่ใช่")
        
        if response:
            # ตรวจสอบคำตอบ
            response_lower = response.lower()
            if any(word in response_lower for word in ['ใช่', 'yes', 'ถูกต้อง', 'ถูก']):
                return True
            elif any(word in response_lower for word in ['ไม่ใช่', 'no', 'ไม่ถูก', 'ผิด']):
                return False
        
        return False
    
    def route_call(self, extension):
        """โอนสายไปยังหมายเลขภายใน"""
        self.log(f"Transferring call to extension {extension}")
        return self.agi_command(f"DIAL SIP/{extension},30")
    
    def main(self):
        """ฟังก์ชันหลักของ AGI script"""
        try:
            self.log("Call routing AGI started")
            
            # 1. รับข้อมูลสาขาที่ต้องการติดต่อ
            branch_input = self.get_speech_input("กรุณาพูดสาขาที่ต้องการติดต่อ")
            
            if not branch_input:
                self.say_text("ขออภัย ไม่สามารถรับข้อมูลได้ กรุณาลองใหม่อีกครั้ง")
                return
            
            self.log(f"Customer requested branch: {branch_input}")
            
            # 2. ค้นหาสาขาในฐานข้อมูล
            branch_info = self.database.get_branch_extension(branch_input)
            
            if not branch_info:
                self.say_text(f"ขออภัย ไม่พบสาขา {branch_input} กรุณาติดต่อเจ้าหน้าที่")
                return
            
            # 3. ยืนยันสาขา
            if not self.confirm_branch(branch_info['branch_name']):
                self.say_text("กรุณาพูดสาขาที่ต้องการติดต่ออีกครั้ง")
                return
            
            # 4. โอนสาย
            self.say_text(f"กำลังโอนสายไปยังสาขา {branch_info['branch_name']}")
            transfer_result = self.route_call(branch_info['extension'])
            
            if transfer_result and "200" in transfer_result:
                self.log("Call transferred successfully")
            else:
                self.say_text("ขออภัย ไม่สามารถโอนสายได้ในขณะนี้ กรุณาติดต่อเจ้าหน้าที่")
            
        except Exception as e:
            self.log(f"Error in main: {e}")
            self.say_text("ขออภัย เกิดข้อผิดพลาดในระบบ กรุณาติดต่อเจ้าหน้าที่")
        
        finally:
            # ปิดการเชื่อมต่อฐานข้อมูล
            self.database.close()

if __name__ == "__main__":
    agi = CallRoutingAGI()
    agi.main()